feat: document user-event v14 (#980)
testing-library/testing-library-docs: docs/user-event/api-pointer.mdx

---
id: pointer
title: Pointer
---

```ts
pointer(input: PointerInput): Promise<void>
```

The `pointer` API allows to simulate interactions with pointer devices. It
accepts a single pointer action or an array of them.

```ts
type PointerInput = PointerActionInput | Array<PointerActionInput>
```

> Our primary target audience tests per `jest` in a `jsdom` environment and there
is no layout in `jsdom`. This means that different from your browser the
elements don't exist in a specific position, layer and size.  
We don't try to determine if the pointer action you describe is possible at that

> position in your layout.

Author: No this literally means that there is no specific position, layer, or size. Elements can have IDL properties and/or CSS styles that usually correspond to elements being at a specific position, having specific dimensions, or being above/under each other. But `jsdom` has no layout. The properties and styles exist, but they have no effect.

```js
document.body.innerHTML=`<img style="position: relative; left: 50px;"/>`
$("img").style.left // '50px'
$("img").offsetLeft // 0
```

> position in your layout.

Reviewer: 👍 Got it, the comment made it click to me - would it be worth it to also mention it in the docs?

> every character has effective width zero and is at effective offset 0,0. (While there might be properties applied that suggest otherwise but have no effect.)

> position in your layout.

Author: I don't know. I feel like it might be better to write a blog post about this topic explaining the concept and then adding a link. Really explaining this, and what difference it makes to testing might be worth it, but is out of scope for these docs.

## Pointer action

There are to types of actions: press and move.

### <a name="press" href="#"/>Pressing a button or touching the screen

A pointer action is a press action if it defines a key to be pressed, to be
released, or both.

```js
pointer({keys: '[MouseLeft]'})
```

You can declare multiple press actions (on the same position) at once which will
be resolved to multiple actions internally. If you don't need to declare any
other properties you can also just supply the `keys` string.

```js
pointer({keys: '[MouseLeft][MouseRight]'})
// or
pointer('[MouseLeft][MouseRight]')
```

> For pressing a button without releasing it, the button name is suffixed with

Reviewer 2: ```suggestion
In order to press a button without releasing it, the button name is suffixed with
```

`>`.  
For just releasing a previously pressed button, the tag is started with `/`.

```js
pointer('[MouseLeft>]') // press the left mouse button
pointer('[/MouseLeft]') // release the left mouse button
```

> Which button are available depends on the [`pointerMap`](options#pointermap).

Reviewer 2: ```suggestion
Which buttons are available depends on the [`pointerMap`](options#pointermap).
```

### <a name="move" href="#"/>Moving a pointer

Every pointer action that is not a press action describes a pointer movement.

You can declare which pointer is moved per `pointerName` property. This defaults
to `mouse`.

Note that the `mouse` pointer (`pointerId: 1`) is also the only pointer that
always exists and has a position. A `touch` pointer only exists while the screen

> is touched and receives a new `pointerId` every time. For these we therefore use

Reviewer 2: ```suggestion
is touched and receives a new `pointerId` every time. For these use cases, we use
```

Author: ```suggestion
is touched and receives a new `pointerId` every time. For these pointers, we use
```

  {pointerName: 'TouchA', target: element2},
  // release the touch pointer at the last position (element2)
  {keys: '[/TouchA]'},
])
```

## Pointer position

### PointerTarget

```ts
interface PointerTarget {
  target: Element
  coords?: PointerCoords
}
```

The `PointerTarget` props allows to describe the position of the pointer on the
document.  
The `coords` you provide are applied as-is to the resulting
[`MouseEvent`](https://developer.mozilla.org/en-US/docs/Web/API/MouseEvent) and
can be omitted.  
The `target` should be the element receiving the pointer input in the browser.
This is the topmost element that can receive pointer event at those coordinates.

### SelectionTarget

```ts
interface SelectionTarget {
  node?: Node
  offset?: number
}
```

Pointer actions can alter the selection in the document.  
In the browser every pointer position corresponds with a DOM position. This is a
DOM node and a DOM offset which usually translates to the character closest to
the pointer position.  
As all character in a no-layout environment are in the same layout position we
assume a pointer position to be closest to the last descendant of the pointer

> `target`.

Reviewer 2: I guess that it's related to the fact that JSDOM has no layout mechanism so all elements have the same position?

> `target`.

Author: Why should we add a comma here? :thinking: 

> As all character in a no-layout environment are in the same layout position

There is no layout. So every character has *effective* width zero and is at *effective* offset `0,0`. (While there might be properties applied that suggest otherwise but have no effect.)

> `target`.

Reviewer: 👍 I see thanks for the elaboration, the term "no-layout" wasn't clear to me.
As for the comma, it's to make it easier to read imho. It's also something that grammarly (a spelling tool) suggests.

> `target`.

Author: I'm no linguistic expert :shrug: 
If I understood some articles about this correctly, generally an introductory phrase should have a comma. But it can be omitted for prepositional phrases shorter than five words and should be omitted if there is no natural pause for building suspension and/or the following sentence could not stand on its own.
I'd judge `Every pointer position corresponds with a DOM position.` on its own as technically incorrect, as this paragraph just explains that this is not the case in the no-layout environment. Following this, I think it should be there, but again, I'm no expert on this. (Which might just have been highlighted by the fact that muscle-memory typed `export` instead of `expert` twice in this post :laughing: )

> `target`.

Reviewer: Yea, me neither 😅
I just added it because I had a corresponding question that went with it.
Thanks again @ph-fritsche !

> If you provide `offset`, we assume the pointer position to be closest to the

Reviewer 2: It seems to me like an example will be good here to make our users understand it properly.

Author: Added an example. I hope this makes it clear enough. Might be worth explaining Layout Positions and DOM offsets and how the browser handles selection per pointer in a blog post, but this would break the mold of docs specific to the library.
